server/server.go

```diff
@@ -2032,8 +2032,14 @@ func (s *Server) loadModules(ctx context.Context, txn storage.Transaction) (map[
 	return modules, nil
 }
 
-func (s *Server) getCompiler() *ast.Compiler {
-	return s.manager.GetCompiler()
+func (s *Server) getCompilerNoChecks() *ast.Compiler {
+	// Strip off any path conflict check as it likely uses an old transaction
+	return s.manager.GetCompiler().WithPathConflictsCheck(nil)
+}
+
+func (s *Server) getCompiler(ctx context.Context, txn storage.Transaction) *ast.Compiler {
+	// Update the path conflict check to use the current context and transaction
+	return s.manager.GetCompiler().WithPathConflictsCheck(storage.NonEmpty(ctx, s.store, txn))
```

Comment:
These calls are going to mutate the shared compiler. I'm a bit worried about this since I can't convince myself that this will not cause any problems.

Reply:
I'm also wondering why the `getCompiler()` call has to be used for read-only queries. I would have assumed those do need path checking to occur. Is it because of partial evaluation? Is the lazy partial evaluation in the server modifying the shared compiler? If so, that feels like an issue on its own. I'm wondering if that lazy partial evaluation is even a good idea--it's useful for development/debug purposes but for production usage, I'd be surprised if people want to potentially introduce that much latency into the request path.

Reply:
> These calls are going to mutate the shared compiler

🤔 💭 💡 🤦‍♂ 

Definitely need to rethink this. Let me take a look at some of the usages... I agree that we shouldn't really need to set it on the majority of the places where we call the helper. I looked around briefly but didn't do a thorough audit. The more I think about it the more I'm convinced there are only a very small number of places where we need the checker.